src/compiler/types.ts

```diff
@@ -603,15 +603,75 @@ namespace ts {
 
     export type DeclarationName = Identifier | StringLiteral | NumericLiteral | ComputedPropertyName | BindingPattern;
 
-    export interface RealDeclaration extends Node {
+    export interface DeclarationBase extends Node {
         _declarationBrand: any;
         name?: DeclarationName;
     }
 
-    // Binary expressions can be declarations if they are 'exports.foo = bar' expressions in JS files
-    export type Declaration = RealDeclaration | BinaryExpression;
-
-    export interface DeclarationStatement extends RealDeclaration, Statement {
+    export type Declaration =

```

Comment:
How does this large discriminated union affect the compile time of the compiler itself?